system/lib/wasmfs/syscalls.cpp

```diff
@@ -719,4 +719,130 @@ long __syscall_getdents64(long fd, long dirp, long count) {
 
   return bytesRead;
 }
+
+long __syscall_rename(long old_path, long new_path) {
+  // The rename syscall must be atomic to prevent other file system operations
+  // from concurrently changing the directories. If it were not atomic, then
+  // other threads could unlink the destination directory after we have moved
+  // our source child. This would leave the file system in an undefined state as
```

Comment:
I'm not sure this is the best example. Unlinking the new path, and orphaning the file, is not that bad: imagine that the unlink happened right after the rename, then the result would be the same. That the file is moved to an orphaned directory is still a consistent state, I think.

An example of actually inconsistent state would be if the file ends up in two places, or has two parents or zero parents. That is, inconsistent state is something that a "filesystem validator" could flag as being obviously wrong. But a file that is unlinked is valid.

Reply:
I remember you provided an example earlier. I included here. Wondering if it works here @kripken ?